### PR TITLE
Fix #128

### LINKED ISSUE #128: Inconsistency with book (Rasmussen+Williams) regarding calculation of std and cov
The current calculation of std and covariance does not follow the book by Rasmussen and Williams. 
Current code follows old code from sklearn, which has since been changed.
For the original report of the problem in sklearn: https://github.com/scikit-learn/scikit-learn/issues/19936
first solution in sklearn: https://github.com/scikit-learn/scikit-learn/pull/19939
later improvement in sklearn: https://github.com/scikit-learn/scikit-learn/pull/19952

Suggestion: rewrite parts of .fit() and .predict() methods of gpr.py to follow style from sklearn

### PR DESCRIPTION
**Pull request by sqbl**
Rephrased math in gaussian processor.

Plot from real data before change:
image

Plot from real data after change:
image

### CODE REVIEW

File: ProcessOptimizer/learning/gaussian_process/gpr.py [hunk 1]
```diff
@@ -327,33 +325,65 @@ def predict(self, X, return_std=False, return_cov=False,
 
         else:  # Predict based on GP posterior
             K_trans = self.kernel_(X, self.X_train_)
-            y_mean = K_trans.dot(self.alpha_)    # Line 4 (y_mean = f_star)
+            y_mean = K_trans @ self.alpha_    # Line 4 (y_mean = f_star)
             # undo normalisation
             y_mean = self.y_train_std_ * y_mean + self.y_train_mean_
```
Comment (dk-teknologisk-mon) [inline ProcessOptimizer/learning/gaussian_process/gpr.py:330]: Over in scikit-learn ([here](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/gaussian_process/_gpr.py)), this line reads:
`y_mean = self._y_train_std * y_mean + self._y_train_mean`
Note the lack of underscores at the end of each variable. If I remember correctly, the values stored in `self._y_train_mean` represent the values computed from the original data, while the value `self._y_train_mean_` (with the underscore) represent the value in the model after fitting? The correct way to undo the normalization is to transform back using the original mean and std, so I just want to make sure that this is what we actually do.

Comment (dk-teknologisk-mon) [reply ProcessOptimizer/learning/gaussian_process/gpr.py:330]: Or is this maybe just the consequence of calling `super(GaussianProcessRegressor, self).fit(X, y)`?

**Review by dk-teknologisk-mon — COMMENTED** [review]
Aside from a comment on line 328, everything else aligns with implementation in scikit-learn and looks good to me :) Once we're sure the above implementation does the right thing it should be good to merge.

**Comment by sqbl** [comment]
Comment from [dk-teknologisk-mon](https://github.com/dk-teknologisk-mon) made me look trhough the implementation again - to realize that the "local" version of y_train_std_ was in fact not consistently used while undoing the normalizations. This is now corrected.
The reason for having variable with and without underscores in the front, is because we still support sklearn 0.24.2. When we skip that support, we ought to be able to make a more lean implementation of the fitting and predicting.

File: ProcessOptimizer/learning/gaussian_process/gpr.py (hunk 1 above)
Comment (sqbl) [reply ProcessOptimizer/learning/gaussian_process/gpr.py:330]: Yes, It is because we leave most parts of the math to sklearns own implementation of the GPR. But as long as we support old sklearn-versions, we have this code to ensure that normalization of y-space is done correctly.

Comment (dk-teknologisk-mon) [reply ProcessOptimizer/learning/gaussian_process/gpr.py:330]: Right, so just to be 100 % sure that I understand it: The present use of `std_` and `mean_` is actually fully equivalent to the scikit-lean implementation, but in our case `self` is just wrapped an extra time around the object?

**Comment by dk-teknologisk-mon** [comment]
I guess one could argue that the "leanest" implementation of fitting and predicting would be to stop trying to perfectly mirror scikit-learn and instead just build on it as a dependency. I know this would require us to accept a lot of extra "stuff" in the dependencies, but at the same time that package has a much greater userbase and is likely to catch bugs and fix them much sooner than we do in our little project :) Something to consider for the future I guess...

**Review by dk-teknologisk-mon — APPROVED** [review]
After discussions, everything looks good to me.

**Comment by sqbl** [comment]
> 

We are already having sklearn do all the math in the fitting. But we have users with both the old and the new implementation of GPR in sklearn (Sklearn GPR changes drastically during the last updates). So while e.g. Brownie Bee is still using sklearn 0.24.2, we maintain support of that. 
... and that is done with the few lines of extra code in the .fit() method compared to the actual call of sklearn own .fit(). 

Furthermore, I've left a bit of legacy code to calculate covariance (instead of just std as implmented by sklearn). This is done because some acquisition functions (at least according to my memory) needs this.

I actually think that the really relevant discussion would be whether it is sklearn or another supplier of GPR that we want as backend(?)

Many words :-)
I can bump to next version when this is merged. I'm happy to add further to the code or the discussion - otherwise feel free to merge.

edit: In the text above, I wrote "calculate covariance". I mean "return_mean_grad"
